src/target/riscv/riscv.c

```diff
@@ -3096,25 +3088,76 @@ static int riscv_write_phys_memory(struct target *target, target_addr_t phys_add
 	return tt->write_memory(target, phys_address, size, count, buffer);
 }
 
-static int riscv_write_memory(struct target *target, target_addr_t address,
-		uint32_t size, uint32_t count, const uint8_t *buffer)
+static int riscv_rw_memory(struct target *target, target_addr_t address, uint32_t size,
+		uint32_t count, uint8_t *read_buffer, const uint8_t *write_buffer)
 {
+	if (read_buffer && write_buffer) {
+		LOG_TARGET_ERROR(target, "Can't handle both read and write buffer.");
+		return ERROR_FAIL;
+	}
+
+	bool is_write = write_buffer ? true : false;
 	if (count == 0) {
-		LOG_TARGET_WARNING(target, "0-length write to 0x%" TARGET_PRIxADDR, address);
+		LOG_TARGET_WARNING(target, "0-length %s 0x%" TARGET_PRIxADDR,
+				is_write ? "write to" : "read from", address);
 		return ERROR_OK;
 	}
 
-	target_addr_t physical_addr;
-	int result = target->type->virt2phys(target, address, &physical_addr);
-	if (result != ERROR_OK) {
-		LOG_TARGET_ERROR(target, "Address translation failed.");
+	int mmu_enabled;
+	int result = riscv_mmu(target, &mmu_enabled);
+	if (result != ERROR_OK)
 		return result;
-	}
 
+	RISCV_INFO(r);
 	struct target_type *tt = get_target_type(target);
 	if (!tt)
 		return ERROR_FAIL;
-	return tt->write_memory(target, physical_addr, size, count, buffer);
+
+
+	if (!mmu_enabled) {
+		if (is_write)
+			return tt->write_memory(target, address, size, count, write_buffer);
+		else
+			return r->read_memory(target, address, size, count, read_buffer, size);
+	}
+
+	result = check_virt_memory_access(target, address, size, count);
+	if (result != ERROR_OK)
+		return result;
+
+	uint32_t current_count = 0;
+	while (current_count < count) {
+		target_addr_t physical_addr;
+		result = target->type->virt2phys(target, address, &physical_addr);
+		if (result != ERROR_OK) {
+			LOG_TARGET_ERROR(target, "Address translation failed.");
+			return result;
+		}
+
+		uint32_t chunk_count = MIN(count - current_count, (RISCV_PGSIZE - RISCV_PGOFFSET(address)) / size);
```

Comment:
This assumes the "worst case" - pages of size 4 KiB. The pages can however be larger than that (e.g. 2MiB).

I don't mind this limitation, though. Please, just make a TODO comment about that, so that it is very clear to future code readers.

For example:

```suggestion
		/* TODO: For simplicity, this algorithm assumes the worst case - the smallest possible page size,
		 * which is  4 KiB. The algorithm can be improved to detect the real page size, and allow to use larger
		 * memory transfers and avoid extra unnecessary virt2phys address translations. */
		uint32_t chunk_count = MIN(count - current_count, (RISCV_PGSIZE - RISCV_PGOFFSET(address)) / size);
```

Reply:
Thanks. Addressed.